[FLINK-19527][Doc]Flink SQL Getting Started (#14003)
apache/flink: docs/dev/table/sql/gettingStarted.md

<!--
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

* This will be replaced by the TOC
{:toc}

Flink SQL enables SQL developers to design and develop the batch or streaming application without writing the Java, Scala, or any other programming language code. It provides a unified API for both batch and streaming APIs. As a user, you can perform powerful transformations. Flink’s SQL support is based on [Apache Calcite](https://calcite.apache.org/) which implements the SQL standard.

> In addition to the SQL API, Flink also has Table API as well with similar semantics as SQL. Table API is a Language integrated API, where we use a specific programming language to write the queries or call the API. For example, we create the table environment, get a table object, and apply different methods that return table API objects. It supports different languages e.g. Java, Scala, Python. 

Reviewer: Try to avoid "we". It isn't clear who "we" refers to. 

```suggestion
In addition to the SQL API, Flink also has a Table API with similar semantics as SQL. The Table API is a language-integrated API, where users develop in a specific programming language to write the queries or call the API. For example, jobs create a table environment, read a table, and apply different transformations and aggregations, and write the results back to another table. It supports different languages e.g. Java, Scala, Python. 
```

> Flink SQL and Table API are just two ways to write queries that use the same API underneath. It wouldn’t be wrong if we say Table API is a wrapper on top of the streaming API. SQL API is a more descriptive way of writing queries using well-known SQL standards where we usually have ‘select * from Table’ pattern.  Table API query starts with from clause, followed by joins and where clause, and then finally projection or select at the last. SQL API is easy to learn and almost everyone knows it already. All the queries are optimized for efficient execution. We will focus on the Flink SQL API, while you can read more about Table API [here]({{ site.baseurl }}/dev/table/).

Reviewer: > Flink SQL and Table API are just two ways to write queries that use the same API underneath. 

It's not really the same API, right? It's two different api's that give you access to the same runtime. 

> It wouldn’t be wrong if we say Table API is a wrapper on top of the streaming API.

I'm not really sure what you're trying to say here. The table API supports both batch and streaming.

Author: I was referring to the same runtime actually as both of them are high level abstractions to the Datastream/dataset apis. Rephrased it.

### Pre-requisites
You only need to have basic knowledge of SQL to follow along. You will not need to write Java or Scala code or use an IDE.

### Installation

> There are various ways to install the Flink. You can download the source code, compile it, and run it. Another option is to have it running inside the container. Probably the easiest one is to download the binaries and run it locally for experimentation. We assume local installation for the rest of the tutorial. You can start a local cluster using the following command from the installation folder

Reviewer: > There are various ways to install the Flink

Its not *the flink* just Flink. "There are various ways to install Flink". 

> You can download the source code, compile it, and run it.

Compiling flink is difficult, I'd rather not mention that in a getting started guide. 

If we are guiding users towards local installation, how about a link to the downloads page.

 
{% highlight bash %}
./bin/start-cluster.sh
{% endhighlight %}
 
Once the cluster is started, it will also start a web server on [localhost:8081](localhost:8081) to manage settings and monitor the different jobs

### SQL Client
SQL Client is an interactive client to submit SQL queries to Flink and visualize the results. It’s like a query editor for any other database management system where you can write queries using standard SQL. You can start the SQL client from the installation folder as follows

 {% highlight bash %}
./bin/sql-client.sh embedded
 {% endhighlight %}

 

> This will start the SQL client in embedded mode. In the future, a user will have two possibilities of starting the SQL Client CLI either by starting an embedded standalone process or by connecting to a remote SQL Client Gateway. At the moment only the embedded mode is supported. 

Reviewer: ```suggestion
```

For a getting started guide, let's not focus on what Flink can't do :)

### Hello World query
 
Once SQL client, our query editor, is up and running we can start writing SQL queries. These queries will be submitted to the Flink cluster for computation and results will be returned to the SQL client UI. Let's start with printing 'Hello World'. You can print hello world using the following simple query
 
{% highlight sql %}
SELECT 'Hello World';
{% endhighlight %}

> The following example will print the current stamp

Reviewer: ```suggestion
We can also execute a built-in function, the following will print the current timestamp
```

{% highlight sql %}
SELECT CURRENT_TIMESTAMP;
{% endhighlight %}

> `Help;` command is used to see different supported DDL (Data definition language) commands. Furthermore, Flink SQL does support different built-in functions as well. The following query will show all the built-in and user defined functions. 

Reviewer: Shouldn't this go before printing the current timestamp?

{% highlight sql %}
SHOW FUNCTIONS;
{% endhighlight %}
 
## Setting up tables

> Real-world database queries are run against the SQL tables. Although Flink is a stream processing engine, users can define a table on top of the streaming data. As we know, all the Flink data processing pipelines generally have three components - source, compute, sink. 

Reviewer: > As we know,

Don't say as we know, new users might not know

> The source is our input or from where we read the data e.g. a text file, Kafka topic. Then we defined some computations that needed to be performed on those data elements. Finally, the sink defines what to do with the output or where to store the results. A sink can be a console log, another output file, or a Kafka topic. It's similar to a database query where we read data from a table, perform a query on it and then display or store the results. 

Reviewer: > Then we defined some computations that needed to be performed on those data elements.

Focus on the present or future tense. We haven't done these things yet. 

> We then define queries to be executed on the rows of that input table.

In Flink SQL semantics, source and sink will be tables, but we know Flink isn’t a storage engine and cannot store the data. So we need to back our table with a [storage connector]({{ site.baseurl }}/dev/table/connect.html) like [file system]({{ site.baseurl }}/dev/table/connect.html#file-system-connector), [Kafka]({{ site.baseurl }}/dev/table/connect.html#kafka-connector) or [MySQL]({{ site.baseurl }}/dev/table/connect.html#jdbc-connector). When we will be defining these tables, we will configure the storage connector type, [format]({{ site.baseurl }}/dev/table/connect.html#table-formats) and schema for each table. 

 
### Input Source Tables

> SQL API environment is configured using the YAML[yaml.org] configuration files. When we start SQL client, it reads the default configuration from the `/conf/sql-client-defaults.yaml` but it can be overriden by user defined configuration file. These files are used to define different environment variables including table source, sinks, [catalogs](({{ site.baseurl }}/dev/table/catalogs.html)), [user-defined functions](({{ site.baseurl }}/dev/table/functions/udfs.html)).

Reviewer: > SQL API environment is configured using the YAML[yaml.org]

Did you mean for this to be al ink?

Reviewer: > When we start SQL client

When we start the SQL client. Unlike Flink, the sql client is a generic term.

Following is an example to define a source table using file system with [csv format]({{ site.baseurl }}/dev/table/connectors/formats/csv.html) in the environment config file but Flink community has added the support for quite a few [formats]({{ site.baseurl }}/dev/table/connectors/formats/). 

{% highlight yaml %}

tables:
  - name: EmployeeTableSource
    type: source-table
    update-mode: append
    connector:
      type: filesystem
      path: "/path/to/something.csv"
    format:
      type: csv
      fields:
        - name: EmpId
          data-type: INT
        - name: EmpName
          data-type: VARCHAR
        - name: DeptId
          data-type: INT          
      line-delimiter: "\n"
      comment-prefix: "#"
    schema:
      - name: EmpId
        data-type: INT
      - name: EmpName
        data-type: VARCHAR
      - name: DeptId
        data-type: INT
{% endhighlight %} 

Addittionally, we can use SQL DDL to [create]({{ site.baseurl }}/dev/table/sql/create.html), [alter]({{ site.baseurl }}/dev/table/sql/alter.html), [drop]({{ site.baseurl }}/dev/table/sql/drop.html) tables from the SQL client. Same table can be defined using DDL as follows

{% highlight sql %}

> CREATE TABLE EmployeeTableSource (

Reviewer: Can we reorder this, lets show off create table first and then mention  yaml as an alternative.

 {% highlight bash %}
./bin/sql-client.sh embedded
 {% endhighlight %} 

### Hello World query
 
Once the SQL client, our query editor, is up and running it's time to start writing SQL queries. These queries will be submitted to Flink cluster for computation and results will be returned to the SQL client UI. Let's start with printing 'Hello World'. You can print hello world using the following simple query
 
{% highlight sql %}
SELECT 'Hello World';
{% endhighlight %}

`Help;` command is used to see different supported DDL (Data definition language) commands. Furthermore, Flink SQL does support different built-in functions as well. The following query will show all the built-in and user-defined functions. 
{% highlight sql %}
SHOW FUNCTIONS;
{% endhighlight %}

Flink SQL provides users with a set of [built-in functions]({{ site.baseurl }}/dev/table/functions/systemFunctions.html) for data transformations. The following example will print the current timestamp using the `CURRENT_TIMESTAMP` function.

{% highlight sql %}
SELECT CURRENT_TIMESTAMP;
{% endhighlight %}

---------------

> {% top %}
> ## Setting up tables

Reviewer: ```suggestion
{% top %}

## Setting up tables
```

Real-world database queries are run against the SQL tables. Although Flink is a stream processing engine, users can define a table on top of the streaming data. Generally, Flink data processing pipelines have three components - source, compute, sink. 

The source is input or from where data is read e.g. a text file, Kafka topic. Then we define some computations that need to be performed on input data. Finally, the sink defines what to do with the output or where to store the results. A sink can be a console log, another output file, or a Kafka topic. It's similar to a database query that reads data from a table, performs a query on it, and then displays the results. 

> In Flink SQL semantics, source and sink will be tables, but Flink isn’t a storage engine hence it cannot store the data. So Flink tables need to backed up with a [storage connector]({{ site.baseurl }}/dev/table/connect.html) like [file system]({{ site.baseurl }}/dev/table/connect.html#file-system-connector), [Kafka]({{ site.baseurl }}/dev/table/connect.html#kafka-connector) or [MySQL]({{ site.baseurl }}/dev/table/connect.html#jdbc-connector). While creating these tables, storage connector type, [format]({{ site.baseurl }}/dev/table/connect.html#table-formats) and schema for each table needs to be defined. 

Reviewer: This is more of a nit but we're trying to migrate links to a new style that enforces correctness. So for example

`{{site.baseurl }}/dev/table/connect.html` becomes {% link /dev/table/connect.md %}`

Author: Should I update all the links?

Aggregations on the continuous stream need to store aggregated results continuously during the execution of the query. For example, if the number of employees needs to be counted for each department from the data stream, the aggregated count needs to be stored for each department during the execution of the query. 

 {% highlight sql %}
 SELECT count(1) as EmpCount, DeptId from EmployeeTableSource GROUP BY DeptId;
 {% endhighlight %} 

Such queries or operators where the state needs to store are referred to as materialized operators. A materialized view caches the result of the query such that the query does not need to be evaluated when the view is accessed. During the execution of the query, the state can grow with every new record processed. 

## Detached mode
So far, We have seen the query execution using the SQL client but does that mean that the SQL client needs to be running all the time? the answer is no. SQL `INSERT INTO` can be used to submit detached queries to Flink cluster. 

 {% highlight sql %}
 INSERT INTO MyTableSink SELECT * FROM MyTableSource
 {% endhighlight %} 
 

> Once submitted such a query, can only be controlled from Flink cluster.  It will run and store the results into the sink table directly instead of loading the results into the system memory. 

Reviewer: It would be nice to have some sort of conclusion and link to some places in the docs as next steps.

Author: I have added help and resources to learn section. 

![image](https://user-images.githubusercontent.com/4470055/100016472-c0290900-2dd9-11eb-8d7a-1276dc83044e.png)

---

> title: "Getting Started - Flink SQL"

Reviewer 2: ```suggestion
title: "Getting Started"
```

Reviewer 2: This will fall under the "SQL" tab, so it could be shortened?

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
-->

* This will be replaced by the TOC
{:toc}

Flink SQL enables SQL developers to design and develop the batch or streaming application without writing the Java, Scala, or any other programming language code. It provides a unified API for both stream and batch processing. As a user, you can perform powerful transformations. Flink’s SQL support is based on [Apache Calcite](https://calcite.apache.org/) which implements the SQL standard.

In addition to the SQL API, Flink also has a Table API with similar semantics as SQL. The Table API is a language-integrated API, where users develop in a specific programming language to write the queries or call the API. For example, jobs create a table environment, read a table, and apply different transformations and aggregations, and write the results back to another table. It supports different languages e.g. Java, Scala, Python. 
 

> Flink SQL and Table API are just two different ways to write queries that use the same Flink runtime. All the queries are optimized for efficient execution. SQL API is a more descriptive way of writing queries using well-known SQL standards e.g. `select * from Table`. On the other hand, Table API queries start with from clause, followed by joins and where clause, and then finally projection or select at the last e.g. `Table.filter(...).select(...)`. Standard SQL is easy and quick to learn even for users with no programming background. This article will focus on Flink SQL API but Table API details can be found [here]({{ site.baseurl }}/dev/table/).

Reviewer 2: I think this would be better placed as an extension to the "Overview" page, otherwise we'll end up with too many core definitions in too many places.

Reviewer 2: What's your opinion on this, @sjwiesman @aljoscha ?

> Flink SQL and Table API are just two different ways to write queries that use the same Flink runtime. All the queries are optimized for efficient execution. SQL API is a more descriptive way of writing queries using well-known SQL standards e.g. `select * from Table`. On the other hand, Table API queries start with from clause, followed by joins and where clause, and then finally projection or select at the last e.g. `Table.filter(...).select(...)`. Standard SQL is easy and quick to learn even for users with no programming background. This article will focus on Flink SQL API but Table API details can be found [here]({{ site.baseurl }}/dev/table/).

Reviewer: +1

> Flink SQL and Table API are just two different ways to write queries that use the same Flink runtime. All the queries are optimized for efficient execution. SQL API is a more descriptive way of writing queries using well-known SQL standards e.g. `select * from Table`. On the other hand, Table API queries start with from clause, followed by joins and where clause, and then finally projection or select at the last e.g. `Table.filter(...).select(...)`. Standard SQL is easy and quick to learn even for users with no programming background. This article will focus on Flink SQL API but Table API details can be found [here]({{ site.baseurl }}/dev/table/).

Author: Took out the initial introduction and added an intro about the tutorial.

### Pre-requisites
You only need to have basic knowledge of SQL to follow along. You will not need to write Java or Scala code or use an IDE.

### Installation

> There are various ways to [install]({{ site.baseurl }}/ops/deployment/) Flink. Probably the easiest one is to download the binaries and run them locally for experimentation. We assume [local installation]({{ site.baseurl }}/try-flink/local_installation.html) for the rest of the tutorial. You can start a local cluster using the following command from the installation folder

Reviewer 2: ```suggestion
There are multiple ways to [install]({{ site.baseurl }}/ops/deployment/) Flink. For experimentation, the most common option is to download the binaries and run them locally. You can follow the steps in ["Local Installation"] to set up the environment ({{ site.baseurl }}/try-flink/local_installation.html) for the rest of the tutorial. 

Once you're all set, use the following command to start a local cluster from the installation folder:
```

Reviewer 2: Don't think it's relevant to link to deployment modes here — this could be confusing to new users.

 
### Input Source Tables
SQL API environment is configured via [YAML](https://yaml.org) configuration files. When the SQL client starts, it reads the default configuration from the `/conf/sql-client-defaults.yaml` but it can be overriden by user defined configuration file. These files are used to define different environment variables including table source, sinks, [catalogs]({{ site.baseurl }}/dev/table/catalogs.html), [user-defined functions]({{ site.baseurl }}/dev/table/functions/udfs.html).

Tables can be defined through the SQL client or using environment config file. The SQL client support [SQL DDL commands]({{ site.baseurl }}/dev/table/sql) similar to traditional SQL. Standard SQL DDL is used to [create]({{ site.baseurl }}/dev/table/sql/create.html), [alter]({{ site.baseurl }}/dev/table/sql/alter.html), [drop]({{ site.baseurl }}/dev/table/sql/drop.html) tables. 

> Flink has a support for few [formats]({{ site.baseurl }}/dev/table/connectors/formats/) that can be used with tables. Following is an example to define a source table on [csv file]({{ site.baseurl }}/dev/table/connectors/formats/csv.html) using DDL with `EmpId, EmpName, DeptId` as header.

Reviewer 2: ```suggestion
Flink has support for a few [formats]({{ site.baseurl }}/dev/table/connectors/formats/) that can be used with tables. Following is an example to define a source table on [csv file]({{ site.baseurl }}/dev/table/connectors/formats/csv.html) using DDL with `EmpId, EmpName, DeptId` as header.
```

> Flink has a support for few [formats]({{ site.baseurl }}/dev/table/connectors/formats/) that can be used with tables. Following is an example to define a source table on [csv file]({{ site.baseurl }}/dev/table/connectors/formats/csv.html) using DDL with `EmpId, EmpName, DeptId` as header.

Reviewer 2: ```suggestion
Flink has a support for different [connectors]({% link dev/table/connectors.md %}) and [formats]({{ site.baseurl }}/dev/table/connectors/formats/) that can be used with tables. Following is an example to define a source table backed by a [CSV file]({{ site.baseurl }}/dev/table/connectors/formats/csv.html) with `EmpId, EmpName, DeptId` as header fields, using SQL DDL.
```

{% highlight sql %}
CREATE TABLE EmployeeTableSource (
    EmpId INT,
    EmpName VARCHAR,
    DeptId INT
) WITH ( 
    'connector' = 'filesystem',
    'path' = '/path/to/something.csv',
    'format' = 'csv'
);
{% endhighlight %} 

> Following is an example to define a source table using file system with [csv format]({{ site.baseurl }}/dev/table/connectors/formats/csv.html) in the environment config file but Flink community has added the support for quite a few [formats]({{ site.baseurl }}/dev/table/connectors/formats/). 

Reviewer 2: ```suggestion
You can also choose to define this source table in the environment configuration file:
```

{% highlight yaml %}

tables:
  - name: EmployeeTableSource
    type: source-table
    update-mode: append
    connector:
      type: filesystem
      path: "/path/to/something.csv"
    format:
      type: csv
      fields:
        - name: EmpId
          data-type: INT
        - name: EmpName
          data-type: VARCHAR
        - name: DeptId
          data-type: INT          
      line-delimiter: "\n"
      comment-prefix: "#"
    schema:
      - name: EmpId
        data-type: INT
      - name: EmpName
        data-type: VARCHAR
      - name: DeptId
        data-type: INT
{% endhighlight %} 

> Now if all employees who belong to a certain department needs to be filtered, then where by clause will be used to filter records on DeptId as follows

Reviewer 2: ```suggestion
If, for example, you need to filter all the employees who belong to a certain department, you can use the `WHERE` clause to filter the records on `DeptId` as follows:
```

{% highlight sql %}
SELECT * from EmployeeTableSource WHERE DeptId = 1;
{% endhighlight %} 

### Output Tables

> The SQL client provides us the output in real-time, but it’s view only. If results need to be stored, a sink or output table where the results of the query will be stored should be defined. Similar to input source tables, sink tables are defined in the environment file or using SQL DDL statements. See the [connection page]({{ site.baseurl }}/dev/table/connect.html) for more information about supported external systems and their configuration.  

Reviewer 2: ```suggestion
The SQL client provides us the output in real-time, but it’s view only. If results need to be stored, a sink or output table where the results of the query will be stored should be defined. Similar to input source tables, sink tables are defined in the environment file or using SQL DDL statements. See ["Table & SQL Connectors"]({{ site.baseurl }}/dev/table/connect.html) for more information about supported external systems and their configuration.  
```

On the other hand, Flink can process data in motion or a continuous stream of data, known as stream processing. In the case of stream processing, query input is unbounded so the result is never final and continuously updated. Having said that, SQL API has different [execution mode](TODO.com) for both batch and streaming processing. However, the semantics of the query is the same, irrespective of whether it is executed one-time on table snapshot or continuously on a changing table. 

Most of the standard SQL operations are available in both batch and streaming mode. Streaming mode has few additional operators which are specific to continuous data processing such as [Temporal tables]({{ site.baseurl }}/dev/table/streaming/temporal_tables.html), [continuous joins queries]({{ site.baseurl }}/dev/table/streaming/joins.html) and [Pattern recognition]({{ site.baseurl }}/dev/table/streaming/match_recognize.html).
<!-- #### Processing mode (append, changelog, ..)-->

> ### SQL Client result modes

Author: I have taken the section out and linked it to the doc section as suggested.

Standard SQL has not been designed with streaming data in mind. As a consequence, there are few conceptual gaps between standard SQL and stream processing. SQL is designed for finite or data-at-rest processing. SQL queries are completed after processing the fixed size of the input data. On the contrary, streaming queries are designed for an infinite amount of data. Therefore, they continue execution forever as queries do not have access to all the data. 

A [continuous query](/table/streaming/dynamic_tables.html#continuous-queries) never terminates and produces a dynamic table as result. [Dynamic tables](/table/streaming/dynamic_tables.html) are the core concept of Flink’s Table API and SQL support for streaming data. 

Aggregations on the continuous stream need to store aggregated results continuously during the execution of the query. For example, if the number of employees needs to be counted for each department from the data stream, the aggregated count needs to be stored for each department during the execution of the query. 

 {% highlight sql %}
 SELECT count(1) as EmpCount, DeptId from EmployeeTableSource GROUP BY DeptId;
 {% endhighlight %} 

> Such queries or operators where the state needs to store are referred to as materialized operators. A materialized view caches the result of the query such that the query does not need to be evaluated when the view is accessed. During the execution of the query, the state can grow with every new record processed. 

Reviewer 2: ```suggestion
Such queries, where state needs to be stored, are referred to as _stateful_. Under the hood, this is similar to a materialized view that caches the result of the query such that the query doesn't need to be evaluated when the view is accessed. During the execution of the query, the state can grow with every new record processed. 
```

Author: @morsapaes  Updated the comments but do we have any suggestion such that if we can link the materialized view?

## Detached mode
So far, We have seen the query execution using the SQL client but does that mean that the SQL client needs to be running all the time? the answer is no. SQL `INSERT INTO` can be used to submit detached queries to Flink cluster. 

 {% highlight sql %}
 INSERT INTO MyTableSink SELECT * FROM MyTableSource
 {% endhighlight %} 
 

> Once submitted such a query, can only be controlled from Flink cluster.  It will run and store the results into the sink table directly instead of loading the results into the system memory. 

Reviewer 2: ```suggestion
Once submitted, such a query can only be controlled from the Flink cluster (for example through the Flink WebUI). It will run and store the results into the sink table directly, instead of loading the results into the system memory. 
```

Author: addressed the inputs except "the Flink"
